Serialization

easyasme opened the pull request:
**Concept**
Historian passes records containing potentally non-JSON-serializable objects to MasterSerializer.
MasterSerializer recursively traverses the record and replaces any objects that needs serialization with dictionary {_ms_type, args, kwargs}.
Dictionary is passed to Blobstorage - simply stores and retrieves these dictionaries without any custom serialization logic - simple implementation.

**Modified**
MasterSerializer - automatically serialize objects by capturing their class name and constructor arguments (args, kwargs).
			for deserialization - it can reconstruct the object by importing the class and calling its constructor with the stored arguments.

PersistenHistory - use MasterSerializer to serialize records before passing them to the blobstorage and to deserialize records after retrieving them

Historian, Blobstorage, workflowmanager - do not need to change

The test file has not been implemented yet; it will be added after confirming that I am on the right track.

byubean commented on .gitignore:
```diff
@@ -5,3 +5,5 @@
 /venv/
 /quest_test/event_tests/.pytest_cache
+poetry.lock
+pyproject.toml
```
this needs to be checked in

byubean commented on demos/guessing_game.py:
```diff
@@ -1,6 +1,9 @@
 from pathlib import Path
+import sys
+
+sys.path.append(str(Path(__file__).resolve().parent / 'src'))
```
Instead, right-click the `src` folder in PyCharm -> Mark Directory As -> Sources Root.
You might also need to do the same for the Quest folder. 

Then remove this code.

byubean commented on src/quest/persistence.py:
```diff
@@ -21,9 +22,10 @@ def delete_blob(self, key: str): ...
 
 class PersistentHistory(History):
-    def __init__(self, namespace: str, storage: BlobStorage):
+    def __init__(self, namespace: str, storage: BlobStorage, masterserializer: MasterSerializer = None):
```
`master_serializer`

byubean commented on src/quest/persistence.py:
```diff
@@ -21,9 +22,10 @@ def delete_blob(self, key: str): ...
+    def __init__(self, namespace: str, storage: BlobStorage, masterserializer: MasterSerializer = None):
         self._namespace = namespace
         self._storage = storage
+        self._masterserializer = masterserializer if masterserializer is not None else MasterSerializer()
```
```py
self._master_serializer = master_serializer or MasterSerializer()
```

byubean replied:
Actually, if the dev isn't passing in a customized MasterSerializer, we should have a dummy default that simply returns the object as-is (to avoid the unnecessary overhead of traversing the data).

byubean commented on src/quest/serializer.py:
```diff
@@ -0,0 +1,65 @@
+        return args, kwargs
+
+    # Imports and returns the class with full module and class name(string)
+    def _import_class(self, full_class_string: str):
+        # split class string -> module path, class name
+        module_path, class_name = full_class_string.rsplit('.', 1)
+        module = importlib.import_module(module_path)
+        # Retrieve class from module
+        cls = getattr(module, class_name)
+        # class object
+        return cls
```
We have this same logic in two places in the code (or will once you and Nic both merge). 

Can you coordinate with Nic to create a single function (maybe in utils?) that can be used wherever it is needed?

byubean commented on src/quest/serializer.py:
```diff
@@ -0,0 +1,65 @@
+            return tuple(self.serialize(v) for v in obj)
+        # Unknown type (Object) - serialize using type and constructor arguments
+        else:
+            args, kwargs = self._get_constructor_args(obj)
+            return {
+                '_ms_type': obj.__module__ + '.' + obj.__class__.__name__,
+                'args': self.serialize(args),
+                'kwargs': self.serialize(kwargs)
+            }
```
My original vision for this was that the user (developer) would register known type serializers (which could be class constructors, but could also be other functions):
```py
class MasterSerializer:
    def __init__(self, type_serializers: dict):
        self._type_serializers = type_serializers
        
    ...
    
    def serialize(self, obj):
        ...
        else:
            factory, args, kwargs = self._type_serializers.get(type(obj))
            return {'_ms_type': get_factory_type(factory), 'args': args, 'kwargs': kwargs }
```
a serializer would look like this:
```py
async def load_message(channel_id, message_id):
    channel = await client.fetch_channel(channel_id)
    message = await channel.fetch_message(message_id)
    return message

def serialize_message(message: Message):
    return load_message, (message.channel.id, message.id), {}

...
master_serializer = MasterSerializer({discord.Message: serialize_message})
```

byubean replied:
You'll have to discern whether a given serializer is async or not (look at `handle_step` for an example). This means that `serialize` and `deserialize` will both have to be async.

byubean replied:
When serializing and deserializing: 
If the factory is a *function*, the name will look different than if it is a *class*. We should support both. 
But if you store the right thing for `_ms_type`, then I think the loading logic should be the same in either case.

easyasme commented:
**Modifications**
MasterSerializer - modified to return the object as-is if no custom serialization is registered.
                          - register_serializer function implemented.
                          - serialize and deserialize method is changed to async.
                          - accepts both class and function.
utils.py - implemented for functions that is used in other files as well.
persistence.py - await serialize and deserialize methods.
manager.py - passes the MasterSerializer when calling _create_history.
__init__.py - _create_history function modified to have MasterSerializer as parameter

**Questions**
persistence.py - I noticed that await can't be used in the __init__ method. I used asyncio.run for the deserialization process. Do you think this is an acceptable solution, or should I refactor the code to use await instead?

serializer.py - In the deserialize function, I think that the factory might be a coroutine function. Do you think I need to handle this case by adding code to support asynchronous factories?

byubean commented on src/quest/persistence.py:
```diff
@@ -35,15 +38,20 @@ def __init__(self, namespace: str, storage: BlobStorage):
         return self._namespace + '.' + md5((item['timestamp'] + item['step_id'] + item['type']).encode()).hexdigest()
 
-    def append(self, item: EventRecord):
+    async def append(self, item: EventRecord):
```
Making this `async` will be an issue. Let's talk this through in class.

byubean commented on src/quest/persistence.py:
```diff
@@ -35,15 +38,20 @@ def __init__(self, namespace: str, storage: BlobStorage):
-                self._items.append(storage.read_blob(key))
+                serialized_item = self._storage.read_blob(key)
+                # Deserialize the item that is read from storage
+                item = asyncio.run(self._master_serializer.deserialize(serialized_item))
```
`asyncio.run` can only be run once in the application (and is typically always at the root). Let's discuss this in class.

easyasme commented:
**Concept**
- Serialization process is initiated by the PersistentHistory.append method which uses MasterSerializer to convert objects.
- Deserialization process occurs when loading workflow history via PersistentHistory.load_history. MasterSerializer is used to restore the serialized objects.
- Historian is aware of the serialization and deserialization process calling methods like load_history, _append_event_record

**Modifications**
Updated Historian, PersistentHistory, workflowmanager, and __init__.py to pass and utilize MasterSerializer

__init__.py: 
- Added the import for MasterSerializer to ensure it's available for the WorkflowManager
-  Modified create_filesystem_manager function to initialize and pass MasterSerializer when creating PersistentHistory
- Adjusted create_history to include master_serializer

manager.py:
- Updated the WorkflowManager to instantiate MasterSerializer
- Modified the _start_workflow method to pass MasterSerializer to both Historian and PersistentHistory

persistence.py:
- load_history method to utilize MasterSerializer for deserialization of stored event records
- Updated the append method to serialize new event records using MasterSerializer
- Updated remove method

**Questions**
- If I could get some feedback on my PR for especially the following files if I'm applying async patterns correctly, or modifying it as the right concept:
   __init__.py, manager.py, historian.py -> load_history, _append_event_record, _record_version_event, handle_step

byubean commented on src/quest/historian.py:
```diff
@@ -11,6 +11,7 @@
 from .history import History
 from .quest_types import ConfigurationRecord, VersionRecord, StepStartRecord, StepEndRecord, \
     ExceptionDetails, ResourceAccessEvent, ResourceEntry, ResourceLifecycleEvent, TaskEvent
+from .serializer import MasterSerializer
```
Import only the protocol

byubean commented on src/quest/__init__.py:
```diff
@@ -10,18 +10,19 @@
     storage = LocalFileSystemBlobStorage(save_folder)
-    history = PersistentHistory(historian_id, storage, master_serializer)
+    history = PersistentHistory(historian_id, storage)
+    serializer = serializer or MasterSerializer()
```
I think we want the default serializer to be the `NoopSerializer`

byubean commented on src/quest/historian.py:
```diff
@@ -186,12 +186,12 @@ def get_function_name(func):
         self.workflow = workflow
         self._configurations: list[tuple[Callable, list, dict]] = []
-        self._master_serializer = master_serializer
+        self._serializer: StepSerializer = serializer or NoopSerializer()
```
I think having the default logic in the `create...` functions makes the most sense, and the Historian simply assigns what it was given.

byubean commented on src/quest/historian.py:
```diff
@@ -468,33 +468,22 @@ def _process_discovered_versions(self):
 
-        # Append the event record using the appropriate method
-        await self._append_event_record(event_record)
-
```
I don't think we want to delete this.

byubean commented on src/quest/historian.py:
```diff
@@ -468,33 +468,22 @@ def _process_discovered_versions(self):
         self._versions[version_name] = version
 
         # Create the VersionRecord
-        event_record = VersionRecord(
```
I don't think this line should change.

byubean commented on src/quest/historian.py:
```diff
@@ -539,24 +528,23 @@ async def handle_step(self, func_name, func: Callable, *args, **kwargs):
             with next_record as record:
                 assert record['step_id'] == step_id, f'{record["step_id"]} != {step_id}'
                 if record['type'] == 'end':
-                    # Rehydrate step from history
-                    assert record['type'] == 'end'
```
Keep this logic

byubean commented on src/quest/historian.py:
```diff
@@ -539,24 +528,23 @@ async def handle_step(self, func_name, func: Callable, *args, **kwargs):
                 if record['type'] == 'end':
-                    # Rehydrate step from history
-                    assert record['type'] == 'end'
+                    # Deserialize result
+                    deserialized_result = await self._serializer.deserialize(record['result'])
```
Move this under the `if` statement on line 533.

byubean commented on src/quest/historian.py:
```diff
@@ -539,24 +528,23 @@ async def handle_step(self, func_name, func: Callable, *args, **kwargs):
                 task_id=self._get_task_name(),
                 step_id=step_id
             )
-            await self._append_event_record(event_record)
```
We need these lines.

byubean commented on src/quest/manager.py:
```diff
@@ -35,7 +35,7 @@ def __init__(self, namespace: str, storage: BlobStorage, create_history: History
         self._workflows: dict[str, Historian] = {}
         self._workflow_tasks: dict[str, asyncio.Task] = {}
-        self._master_serializer = MasterSerializer()
+        self._serializer: StepSerializer = MasterSerializer()
```
The serializer dependency should be provided in the constructor - we shouldn't hard-code `MasterSerializer` here.

byubean commented on src/quest/manager.py:
```diff
@@ -62,17 +62,16 @@ def _remove_workflow(self, workflow_id: str):
-        self._master_serializer.register_serializer(obj_type, serializer_func)
+    def register_serializer(self, obj_type: type, serializer: TypeSerializer[Any]):
+        self._serializer.register_serializer(obj_type, serializer)
 
```
Instead of this function, the developer should now provide the type serializers to the MasterSerializer directly, and then provide the MasterSerializer (or the NoopSerializer if desired) to the WorkflowManager constructor directly.
```py
type_sers = {
  MyClass: lambda my_class: ...,
  YourClass: lambda your_class: ...
}
serializer = MasterSerializer(type_sers)
...
workflow_manager = WorkflowManager(..., serializer)
```

byubean commented on src/quest/serializer.py:
```diff
@@ -1,13 +1,41 @@
-class MasterSerializer:
-    def __init__(self, type_serializers=None):
-        # Initialize with a dictionary of custom serializers - default to empty dict
+class StepSerializer(Protocol):
```
Look at the signatures I put in the scratch file. Note how the generic var `T` is used.

byubean replied:
in fact, looking at the usage below, I think we want something like this:
```py
class TypeFactory(Protocol[T]):
    async def __call__(self, *args, **kwargs) -> T: ...

class TypeSerializer(Protocol[T]):
    async def serialize(self, obj: T) -> tuple[TypeFactory[T], tuple, dict]: ...
```

then below you can delete `_get_serializer_for_factory` and instead just load the factory from the name/module stored in the dictionary.

We may need a check to ensure that the deserializer provided is not a lambda, as you can't dynamically import lambdas. If they try to use a lambda, we should throw a ValueError.

byubean commented on src/quest/serializer.py:
```diff
@@ -1,13 +1,41 @@
+class MasterSerializer(StepSerializer):
+    def __init__(self, type_serializers: Dict[type, TypeSerializer[Any]] = None):
         self._type_serializers = type_serializers or {}
+        self._default_serializer = NoopSerializer()
```
This isn't needed in `MasterSerializer`.

byubean commented on src/quest/serializer.py:
```diff
@@ -1,13 +1,41 @@
         self._type_serializers = type_serializers or {}
+        self._default_serializer = NoopSerializer()
+
+    def register_serializer(self, obj_type: type, serializer: TypeSerializer[Any]):
+        self._type_serializers[obj_type] = serializer
```
I think it's safe to require all type serializers to be provided in the constructor (and thus delete this method)

byubean commented on src/quest/serializer.py:
```diff
@@ -21,19 +49,18 @@ async def serialize(self, obj: Any) -> Any:
-            factory, args, kwargs = serializer(obj)
-            # Get full name of the factory (function, class)
+        serializer = self._get_serializer_for_type(type(obj))
+        if serializer is not self._default_serializer:
```
`if serializers is not None:`

byubean commented on src/quest/serializer.py:
```diff
@@ -21,19 +49,18 @@ async def serialize(self, obj: Any) -> Any:
             return {
                 '_ms_factory': factory_name,
                 'args': await self.serialize(args),
                 'kwargs': await self.serialize(kwargs)
```
I would make a TypedDict for this

byubean commented on src/quest/serializer.py:
```diff
@@ -45,23 +72,41 @@ async def deserialize(self, data: Any) -> Any:
-                return {await self.deserialize(k): await self.deserialize(v) for k, v in data.items()}
+                # Reconstruct object
+                serializer = self._get_serializer_for_factory(factory)
+                if serializer is not self._default_serializer:
```
`if serializer is not None`

byubean commented on src/quest/serializer.py:
```diff
@@ -45,23 +72,41 @@ async def deserialize(self, data: Any) -> Any:
+            if issubclass(obj_type, registered_type):
+                return serializer
+        # Nothing match
+        return self._default_serializer
```
`return None`

byubean commented on src/quest/serializer.py:
```diff
@@ -45,23 +72,41 @@ async def deserialize(self, data: Any) -> Any:
+        # If the object's type registered in self._type_serializers dictionary
+        if obj_type in self._type_serializers:
+            return self._type_serializers[obj_type]
+        # check if obj_type is subclass of registered type
+        for registered_type, serializer in self._type_serializers.items():
+            if issubclass(obj_type, registered_type):
+                return serializer
```
Inheritance is rarely used anymore. I think we can just support direct matches.

byubean commented on src/quest/utils.py:
```diff
@@ -14,5 +14,5 @@ def get_obj_name(obj):
     # Import and return object with its full name
     module_name, attr_name = full_name.rsplit('.', 1)
-    module = importlib.import_module(module_name)
+    module = __import__(module_name, fromlist=[attr_name])
```
I think the `importlib.import_module` syntax is preferred.

byubean commented on src/quest/historian.py:
```diff
@@ -473,13 +475,14 @@ def _record_version_event(self, version_name, version):
 
-        self._history.append(VersionRecord(
+        # Create the VersionRecord
+        VersionRecord(
             type='set_version',
             timestamp=_get_current_timestamp(),
             step_id=version_name,
             task_id=self._get_external_task_name(),  # the external task owns these
             version=version
-        ))
```
We need this too

byubean commented on quest_test/test_serializer.py:
```diff
@@ -0,0 +1,110 @@
+from dataclasses import dataclass
+import pytest
+
+from src.quest.historian import Historian
+from src.quest.serializer import MasterSerializer, StepSerializer
+from src.quest.wrappers import step
+from src.quest.quest_types import EventRecord
```
no `src.`

byubean commented on quest_test/test_serializer.py:
```diff
@@ -0,0 +1,110 @@
+        kwargs = {}
+        return factory, args, kwargs
+
+    async def deserialize(self, data):
+        pass
+
```
```py
async def deserialize(self, *args, **kwargs) -> Stuff:
    return Stuff(*args, **kwargs)
```

byubean commented on quest_test/test_serializer.py:
```diff
@@ -0,0 +1,110 @@
+
+# TypeSerializer for Stuff
+class StuffSerializer:
+    async def serialize(self, obj: Stuff):
```
```py
async def serializer(self, obj: Stuff) -> tuple, dict:
    pass
```
